### PR TITLE
New debug feature: IPC message injector

### PR DESCRIPTION
**Pull request by ujfalusi**
Hi,

With the new SND_SOC_SOF_DEBUG_IPC_MSG_INJECTOR it is going to be possible to inject arbitrary messages via the `debugfs/sof/ipc_msg_inject` file and get the reply from the same file as a binary.
The main use of this feature is to stress test the firmware's robustness against maliciously (or erroneous) crafted messages.
We also receive firmware crash reports with only a binary of the sent message which caused the firmware crash.

Before adding the new debug feature I needed to make sure that the IPC reply handling is working correctly for cases when we don't know beforehand the size of the reply (since we don't know what kind of message we are sending).
While doing this I have noticed that the `memory_info` debugfs file is unusable as the size check in `hda_dsp_ipc_get_reply()` only makes exception for PROBES messages. `memory_info` should not have worked at all on any other platform.
Basically the attempt by 85bfa34a88d6d31c5bbec50d2077f1fbc117b290 broke the `memory_info` for HDA also by accident.

Therefore the first three patch in this PR is adding a generic get_reply implementation to be used by all platforms (their implementation was identical).
I'm aware that PR #3092 will change the sof_mailbox_read() use, but I can rebase easily.

@lgirdwood, is this something you were expecting when asked for similar feature?

### CODE REVIEW

**Comment by keyonjie** [comment]
good work @ujfalusi .
@libinyang FYI. Could be similar to the fuzzy IPC coverage we talked today.

**Comment by libinyang** [comment]
> good work @ujfalusi .
> @libinyang FYI. Could be similar to the fuzzy IPC coverage we talked today.

Yes, it seems a great feature for vulnerability test.

File: sound/soc/sof/Kconfig [hunk 1]
```diff
@@ -188,6 +188,14 @@ config SND_SOC_SOF_DEBUG_IPC_FLOOD_TEST
 	  Say Y if you want to enable IPC flood test.
 	  If unsure, select "N".
 
+config SND_SOC_SOF_DEBUG_IPC_MSG_INJECTOR
+	bool "SOF enable IPC message injector"
+	help
+	  This option enables the IPC message injector which can be used to send
+	  crafted IPC messages to the DSP to test it's robustness.
```
Comment (plbossart) [inline sound/soc/sof/Kconfig]: its

File: sound/soc/sof/Kconfig [hunk 2]
```diff
@@ -188,6 +188,14 @@ config SND_SOC_SOF_DEBUG_IPC_FLOOD_TEST
...
+	  Say Y if you want to enable IPC flood test.
```
Comment (plbossart) [inline sound/soc/sof/Kconfig]: copy-paste? this doesn't look related to the IPC flood test?

File: sound/soc/sof/debug.c [hunk 3]
```diff
@@ -812,6 +914,13 @@ int snd_sof_dbg_init(struct snd_sof_dev *sdev)
 		return err;
 #endif
 
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_DEBUG_IPC_MSG_INJECTOR)
+	err = snd_sof_debugfs_msg_inject_item(sdev, "ipc_msg_inject", 0644,
+					      &msg_inject_fops);
+	if (err < 0)
+		return err;
+#endif
+
```
Comment (plbossart) [inline sound/soc/sof/debug.c:921]: I may be missing something: we are moving the IPC flood test to the new SOF client framework. Why not do the same for the IPC message injector?

And along the same lines. could the IPC flood test use crafted messages? Is there any merit/rationale in keeping the two test capabilities separate?

**Review by plbossart — COMMENTED** [review]
two nit-picks and a more general comment below. Thanks @ujfalusi

File: sound/soc/sof/debug.c (hunk 3 above)
Comment (ujfalusi) [reply sound/soc/sof/debug.c:921]: The SOF clients are not there yet and this feature has been requested with high urgency.

I will be moving it to a client and it is going to be the `sof-client-ipc-test.c`, which is atm in the draft PR is the ipc flood test. My only  concern with it is that we support more than one ipc_flood instance, thus more than one device for the ipc-test.
I guess I can use the same method to create the message injection only for a single device as I do with the symlink for the flood support.

If there is a need later we can add crafted message support to the flood, but it is really not in a scope of neither PR.

File: sound/soc/sof/Kconfig (hunk 1 above)
Comment (ujfalusi) [reply sound/soc/sof/Kconfig]: Certainly!

File: sound/soc/sof/Kconfig (hunk 2 above)
Comment (ujfalusi) [reply sound/soc/sof/Kconfig]: Ah, yes.

**Comment by ujfalusi** [comment]
Changes since v1:
- Fixed grammar and copy paste issues pointed out by @plbossart
- rebased on sof-dev
 - use `snd_sof_dsp_mailbox_read` in `sof_mailbox_get_reply()`

**Review by dbaluta — DISMISSED** [review]

**Comment by plbossart** [comment]
re-running GitHub Actions with https://github.com/thesofproject/kconfig/pull/61

@ujfalusi please double-check https://github.com/thesofproject/kconfig/ to make sure we have all the latest and greatest configs to support your changes, thanks!

**Review by plbossart — APPROVED** [review]

File: sound/soc/sof/debug.c [hunk 4]
```diff
@@ -336,6 +336,108 @@ static int sof_debug_ipc_flood_test(struct snd_sof_dev *sdev,
...
+	size_t size;
+	int ret, err;
+
+	if (!count || *ppos || count > SOF_IPC_MSG_MAX_SIZE)
+		return 0;
+
+	size = simple_write_to_buffer(dfse->msg_inject_tx, count, ppos, buffer,
+				      count);
+
+	if (hdr->size != count) {
```
Comment (ujfalusi) [inline sound/soc/sof/debug.c]: This check should be dropped to allow messages with wrong size to be passed to DSP as well.

**Comment by ujfalusi** [comment]
Changes since v2:
- drop size check for file size vs size in message header
- set the unused part of the tx buffer to 0

These changes going to allow messages with bogus size encoded to be passed tot he firmware.

File: sound/soc/sof/debug.c [hunk 5]
```diff
@@ -812,6 +909,13 @@ int snd_sof_dbg_init(struct snd_sof_dev *sdev)
 		return err;
 #endif
 
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_DEBUG_IPC_MSG_INJECTOR)
+	err = snd_sof_debugfs_msg_inject_item(sdev, "ipc_msg_inject", 0644,
+					      &msg_inject_fops);
+	if (err < 0)
+		return err;
```
Comment (plbossart) [inline sound/soc/sof/debug.c:919]: You may want to duplicate the comment
````
/* errors are only due to memory allocation, not debugfs */
````

Some maintainers insist that we shall not stop the probe when debugfs fails, and in reviews they routinely question error handling if you don't mention what errors can happen

**Review by plbossart — COMMENTED** [review]
minor comment below

File: sound/soc/sof/debug.c (hunk 5 above)
Comment (ujfalusi) [reply sound/soc/sof/debug.c:919]: Oh, thanks for the warning, I'll do that!

File: sound/soc/sof/debug.c [hunk 6]
```diff
@@ -336,6 +336,103 @@ static int sof_debug_ipc_flood_test(struct snd_sof_dev *sdev,
...
+					   const struct file_operations *fops)
+{
+	struct snd_sof_dfsentry *dfse;
+
+	dfse = devm_kzalloc(sdev->dev, sizeof(*dfse), GFP_KERNEL);
+	if (!dfse)
+		return -ENOMEM;
+
+	/* pre allocate the tx and rx buffers */
+	dfse->msg_inject_tx = devm_kzalloc(sdev->dev, SOF_IPC_MSG_MAX_SIZE, GFP_KERNEL);
```
Comment (RanderWang) [inline sound/soc/sof/debug.c:421]: SOF_IPC_MSG_MAX_SIZE is very limited and only works for small size of ipc message. you can use sdev->host_box.size which is the max size of ipc message.  For IPC4 case, we will not use SOF_IPC_MSG_MAX_SIZE for above reason

Comment (RanderWang) [reply sound/soc/sof/debug.c:421]: with such change, you need to refine your inject_write function

Comment (ujfalusi) [reply sound/soc/sof/debug.c:421]: @RanderWang, the IPC core can not handle anything larger than `SOF_IPC_MSG_MAX_SIZE`, because:
https://github.com/thesofproject/linux/blob/5a6839b58e44edb5a0f5811d3573bca26cbf1217/sound/soc/sof/ipc.c#L931
and
https://github.com/thesofproject/linux/blob/5a6839b58e44edb5a0f5811d3573bca26cbf1217/sound/soc/sof/ipc.c#L936

This is a limitation we have at the moment.

File: sound/soc/sof/utils.c [hunk 7]
```diff
@@ -69,6 +69,50 @@ void sof_mailbox_read(struct snd_sof_dev *sdev, u32 offset,
...
+	/* get the generic reply */
+	snd_sof_dsp_mailbox_read(sdev, sdev->host_box.offset, &reply,
+				 sizeof(reply));
+
+	if (reply.error < 0) {
+		memcpy(msg->reply_data, &reply, sizeof(reply));
+		ret = reply.error;
+	} else if (msg->reply_size > 0) {
+		if (reply.hdr.size > msg->reply_size) {
+			dev_err(sdev->dev,
```
Comment (ujfalusi) [inline sound/soc/sof/utils.c]: It looks like that there are cases when firmware will indicate no error in reply but at the same time the size is set to 0 which by itself should be an error as the `reply.hdr.size` _must_ be at least the `sizeof(reply)`

**Comment by ujfalusi** [comment]
Changes since v3:
- Comment: `errors are only due to memory allocation, not debugfs` added after snd_sof_debugfs_msg_inject_item() as recommended by @plbossart 
- The generic `sof_mailbox_get_reply()` now handles replies with `reply.hdr.size = 0` as error
- dev_dbg() print added when the received message is shorter than the provided buffer. This will trigger for probes, mem_info and injector operations but might be useful to catch real issues

File: sound/soc/sof/debug.c (hunk 6 above)
Comment (plbossart) [reply sound/soc/sof/debug.c:421]: Not able to follow your comment @RanderWang, can you elaborate?

File: sound/soc/sof/utils.c (hunk 7 above)
Comment (ranj063) [reply sound/soc/sof/utils.c]: ok, so are you saying you need to fix this? or just fyi @ujfalusi ?

**Review by ranj063 — APPROVED** [review]
looks good to me @ujfalusi

**Review by ranj063 — APPROVED** [review]

File: sound/soc/sof/utils.c (hunk 7 above)
Comment (ujfalusi) [reply sound/soc/sof/utils.c]: @ranj063, it was a comment for myself to point out a bug or regression the v3 would have introduced (I can not ask for changes on my own PR afaik).
I have addressed it in v4.

**Comment by ujfalusi** [comment]
@plbossart, @ranj063, @dbaluta, @kv2019i, I wonder if I should move the `sof_mailbox_get_reply()` from utils.c to ipc.c as `sof_ipc_get_reply()` or `sof_ipc_process_reply()` or something along these lines?
It is using snd_sof_dsp_mailbox_read(), so it is platform agnostic and the rest of the function is just to validate the reply and read it from the mailbox.

**Comment by ranj063** [comment]
> @plbossart, @ranj063, @dbaluta, @kv2019i, I wonder if I should move the `sof_mailbox_get_reply()` from utils.c to ipc.c as `sof_ipc_get_reply()` or `sof_ipc_process_reply()` or something along these lines?
> It is using snd_sof_dsp_mailbox_read(), so it is platform agnostic and the rest of the function is just to validate the reply and read it from the mailbox.

yes, makes sense

File: sound/soc/sof/debug.c (hunk 6 above)
Comment (RanderWang) [reply sound/soc/sof/debug.c:421]: For current sof, we split big ipc msg into small batch of ipc msg, please check sof_set_get_large_ctrl_data, which is for filter data or other byte array.  And for ipc4, the max ipc size will be not SOF_IPC_MSG_MAX_SIZE. I will create a PR to address ipc3 & ipc4.

Comment (RanderWang) [reply sound/soc/sof/debug.c:421]: @ujfalusi  you can ignore ipc4 since I can make change in my PR based on your work, but current large ipc message support should be included

Comment (ujfalusi) [reply sound/soc/sof/debug.c:421]: @RanderWang, the API for `sof_set_get_large_ctrl_data()` is far from a generic interface to be able to pass any sized IPC message and afaik it can only be done for `SOF_IPC_GLB_COMP_MSG` type of messages.
I guess we could try to interpret the injected message which is larger than SOF_IPC_MSG_MAX_SIZE as `struct sof_ipc_ctrl_data` and try to break it down.
This would require validation of the data given to be injected to be able to craft split messages. To complicate this, I have received a binary message which supposed to cause issues on the firmware side, the file is 388 bytes long, so a long message but the data within does not make sense, does not map to `struct sof_ipc_ctrl_data` so we would not be able to pass it anyways.

It might be possible to get it working with valid binary if I can get hold of one...

Comment (RanderWang) [reply sound/soc/sof/debug.c:421]: yes, I know this is a corner case and make it much more complex. I also don't have any experience for such big ipc message. @ranj063 , do you have any idea ?

Comment (ujfalusi) [reply sound/soc/sof/debug.c:421]: One of the example binary message file I have received starts with:
```
00000000  11 00 00 00 2e 01 02 50  38 74 86 13 a3 73 c4 18  |.......P8t...s..|
00000010  93 ee 54 22 03 e7 d7 70  aa db 98 43 a8 30 e8 da  |..T"...p...C.0..|
00000020  52 46 81 24 51 34 32 39  34 39 34 39 36 37 35 35  |RF.$Q42949496755|
00000030  34 c8 07 f1 5f dd 82 da  c9 2d 82 e9 be bc f1 ed  |4..._....-......|
```
The file is 388 bytes long, so it supposed to be a long message. Even the first u32 which must be the size in the header looks suspicious with it's `17`, but the rest is not saner either if I map it to `struct sof_ipc_ctrl_data cdata`:
```
cdata {
	.rhdr {
		.hdr {
			.size = 17;
			.cmd = 0x5002012e;
		};
		.error = 0x13867438;
	};
	.comp_id = 0x18c473a3;
	.type = 0x2254ee93;
	.cmd = 0x70d7e703;
	.index = 0x4398dbaa;
	.buffer {
		.hdr {
			.size = 0xdae830a8;
			.cmd = 0x24814652;
		};
		.phy_addr = 0x39323451
		.pages = 0x39343934;
		.size = 0x35353736;
		.reserved[3];
	};
	.num_elems = 0xedf1bcbe;
	.elems_remaining = 0xd60c29be;
	.msg_index = 0x7ed5ffd4;
....
};
```
It juts makes no sense and can not be parsed, it is supposed to be a `SOF_IPC_GLB_COMP_MSG | SOF_IPC_COMP_GET_VALUE`, but I don't know what it is and how to interpret and handle it in a Linux safe way...

**Comment by ujfalusi** [comment]
Changes since v4:
- the generic reply retriever is moved from utils.c to ipc. as `snd_sof_ipc_get_reply(sdev)` and additional wrapper added to bundle it together with the `snd_sof_ipc_reply()` as `snd_sof_ipc_process_reply(sdev, msg_id)`. This can be used by all platforms except hda.

File: sound/soc/sof/debug.c (hunk 6 above)
Comment (ranj063) [reply sound/soc/sof/debug.c:421]: @RanderWang @ujfalusi I think we should keep this simple like it is now and skip large IPCs. They are specifically for comp data and we already have the cset tlv mechanism if we want to validate large IPCs already.

**Review by ranj063 — DISMISSED** [review]

File: sound/soc/sof/debug.c (hunk 6 above)
Comment (lgirdwood) [reply sound/soc/sof/debug.c:421]: lets not enforce an ABI specific size check in either the header or the write size from userspace. Would be nice to alloc a PAGE and use this as the max size.

**Review by kv2019i — DISMISSED** [review]
Code looks good. We do need a test case for this in sof-test to ensure this also stays working. And yeah, yet another Kconfig option, but we have a Kconfig for each dev option separately and this is following the model.

File: sound/soc/sof/debug.c (hunk 6 above)
Comment (ujfalusi) [reply sound/soc/sof/debug.c:421]: > lets not enforce an ABI specific size check in either the header or the write size from userspace. Would be nice to alloc a PAGE and use this as the max size.

@lgirdwood, just to make sure I understand it right, you are suggesting:
```c
/* pre allocate the tx and rx buffers */
dfse->msg_inject_tx = devm_kzalloc(sdev->dev, PAGE_SIZE, GFP_KERNEL);
dfse->msg_inject_rx = devm_kzalloc(sdev->dev, PAGE_SIZE, GFP_KERNEL);
if (!dfse->msg_inject_tx || !dfse->msg_inject_rx)
	return -ENOMEM;
```

Comment (ujfalusi) [reply sound/soc/sof/debug.c:421]: @lgirdwood, the core is not allowing msg or reply size larger than SOF_IPC_MSG_MAX_SIZE:
https://github.com/thesofproject/linux/blob/eb55816743ecf7a68a94e0306d069c5e50783722/sound/soc/sof/ipc.c#L354-L362

**Review by RanderWang — DISMISSED** [review]
it is enough for 99% case.  I will figure out another path for IPC4

File: sound/soc/sof/ipc.c [hunk 8]
```diff
@@ -373,6 +373,65 @@ int sof_ipc_tx_message_no_pm(struct snd_sof_ipc *ipc, u32 header,
...
+				reply.hdr.size, msg->reply_size);
+			ret = -EINVAL;
+		} else {
+			if (msg->reply_size != reply.hdr.size)
+				dev_dbg(sdev->dev,
+					"expected reply size %zu, got %u bytes\n",
+					msg->reply_size, reply.hdr.size);
+
+			if (msg->reply_size > reply.hdr.size)
+				msg->reply_size = reply.hdr.size;
```
Comment (lyakh) [inline sound/soc/sof/ipc.c]: after the `else` in line 415 conditions `msg->reply_size != reply.hdr.size` and `msg->reply_size > reply.hdr.size` are equivalent, so, it would be better to use the same form to avoid confusion? and you can join the ones in lines 416 and 421

File: sound/soc/sof/debug.c [hunk 9]
```diff
@@ -336,6 +336,103 @@ static int sof_debug_ipc_flood_test(struct snd_sof_dev *sdev,
...
+	struct snd_sof_dev *sdev = dfse->sdev;
+	struct sof_ipc_cmd_hdr *hdr = dfse->msg_inject_tx;
+	size_t size;
+	int ret, err;
+
+	if (!count || *ppos || count > SOF_IPC_MSG_MAX_SIZE)
+		return 0;
+
+	size = simple_write_to_buffer(dfse->msg_inject_tx, count, ppos, buffer,
+				      count);
```
Comment (lyakh) [inline sound/soc/sof/debug.c]: check for `size < 0` or `size != count`?

File: sound/soc/sof/debug.c [hunk 10]
```diff
@@ -336,6 +336,103 @@ static int sof_debug_ipc_flood_test(struct snd_sof_dev *sdev,
...
+	memset(dfse->msg_inject_tx + count, 0, SOF_IPC_MSG_MAX_SIZE - count);
```
Comment (lyakh) [inline sound/soc/sof/debug.c]: having verified, that `size == count`, this shouldn't be needed?

File: sound/soc/sof/ipc.c (hunk 8 above)
Comment (ujfalusi) [reply sound/soc/sof/ipc.c]: @lyakh, right, it is enough to have a single `if (reply.hdr.size < msg->reply_size)` and have the print and the msg->reply_size fixup.

File: sound/soc/sof/debug.c (hunk 10 above)
Comment (ujfalusi) [reply sound/soc/sof/debug.c]: The message itself still can say that `hdr.size > count`, I want to make sure that in this case we _know_ that the bytes after count is always 0.

File: sound/soc/sof/debug.c (hunk 9 above)
Comment (ujfalusi) [reply sound/soc/sof/debug.c]: OK, I can do that.

**Comment by ujfalusi** [comment]
> it is enough for 99% case. I will figure out another path for IPC4

@RanderWang, we can update this very same feature to be universal when the core can cope with different maximums?

**Comment by ujfalusi** [comment]
Changes since v5:
- Simplify the checks in `snd_sof_ipc_get_reply()` as suggested by @lyakh 
- We can drop most of the direct checks in `msg_inject_write()` by leaving it to `simple_write_to_buffer()` and check the return value

File: sound/soc/sof/debug.c (hunk 10 above)
Comment (lyakh) [reply sound/soc/sof/debug.c]: why do you need those bytes as 0? Any specific reason for that?

Comment (ujfalusi) [reply sound/soc/sof/debug.c]: @lyakh, I don't want them to be random. It would be impossible to reproduce and fix a failure if we send random and not recorded values to the firmware.
If we want to test random values in messages then the message should be crafted that way.

**Comment by ujfalusi** [comment]
Changes since v6:
- Optimize the number of comparisons in the generic `snd_sof_ipc_get_reply()` for the most common tx scenario (when `reply.hdr.size == msg->reply_size`)

File: sound/soc/sof/debug.c (hunk 10 above)
Comment (lyakh) [reply sound/soc/sof/debug.c]: sure, no, we shouldn't be sending random data. But are those bytes sent in the IPC? Aren't only `count` bytes sent in lines 388-389 below?

Comment (ujfalusi) [reply sound/soc/sof/debug.c]: @lyakh, aaah, if you put it that way... You are right, the memset is pointless unless we use `SOF_IPC_MSG_MAX_SIZE` in place of count.
We can either drop the memset or make the core to copy the full msg_buffer to mailbox unconditionally.
What do you think?

Comment (lyakh) [reply sound/soc/sof/debug.c]: as long as there's no functional requirement or argument in favour of copying the whole buffer, I'd limit to only the required size.

Comment (ujfalusi) [reply sound/soc/sof/debug.c]: OK, I'll drop the memset for now. It can be added back if we saw it is needed.

Do you have other concerns before I push an update?

**Comment by ujfalusi** [comment]
Changes since v7:
- drop memset for the tx message

**Review by RanderWang — APPROVED** [review]

**Review by ranj063 — APPROVED** [review]

**Comment by ranj063** [comment]
@mengdonglin It will be good to use this new feature to add some more tests to our SOFCI test suite. One thing that comes to my mind right away, is stress testing the secondary core power up/down for multi-core.

**Review by plbossart — APPROVED** [review]

File: sound/soc/sof/ipc.c [hunk 11]
```diff
@@ -373,6 +373,69 @@ int sof_ipc_tx_message_no_pm(struct snd_sof_ipc *ipc, u32 header,
...
+			dev_err(sdev->dev,
+				"reply size (%u) exceeds the buffer size (%zu)\n",
+				reply.hdr.size, msg->reply_size);
+			ret = -EINVAL;
+		}
+
+		/* get the full message if reply.hdr.size <= msg->reply_size */
+		if (!ret)
+			snd_sof_dsp_mailbox_read(sdev, sdev->host_box.offset,
+						 msg->reply_data, msg->reply_size);
```
Comment (lyakh) [inline sound/soc/sof/ipc.c:431]: I understand that it's following the functions that you're replacing, so not an issue for this PR, but why are we then reading the larger size? Above we check `if (reply.hdr.size < msg->reply_size)` (btw, in that case you only print a debug message, wouldn't it be worth printing some warning?) and set `ret = 0` but then we read `msg->reply_data` bytes which can be larger than what the sender has sent? Is there a specific reason behind this? If not, maybe it would be safer to read the smaller amount. E.g. the injector also specifies `SOF_IPC_MSG_MAX_SIZE` as reply size, not because it expects that, but just as a maximum size it can handle. Shouldn't be needed to copy more than the firmware has sent?

**Review by lyakh — APPROVED** [review]

File: sound/soc/sof/ipc.c (hunk 11 above)
Comment (ujfalusi) [reply sound/soc/sof/ipc.c:431]: @lyakh,
if `reply.hdr.size < msg->reply_size` then we update the reply_size: `msg->reply_size = reply.hdr.size` and we print only debug level print as this is fine. Probes does this and also other messages pass max sized reply buffer since they don't know the size which will come back. Then I'm introducing the 'injector' which will certainly have no idea what will come back as it does not know what it is sending.

In short: we already have more uses of reply_size = SOF_IPC_MSG_MAX_SIZE from IPC users, it is not an error. The error case is when more data is received than what the user provided.
